sklearn/preprocessing/_discretization.py

```diff
@@ -242,15 +242,23 @@ def fit(self, X, y=None, sample_weight=None):
             # Take a subsample of `X`
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
+                use_weights_in_resampling=use_weights_in_resampling,
             )
```

Comment:
```suggestion
            )
            # Since we already used the weights when resampling when provided,
            # we set them back to `None` to avoid accounting for the weights twice
            # in subsequent operations to compute weight-aware bin edges with
            # quantiles or k-means.
            sample_weight = None
```